### ISSUE TITLE
Revisit dataset-related parameter names

### ISSUE THREAD

**Issue by AndresTanasijczuk** [dmwm/CRABClient#4559]
1) We use `inputDataset` for the main input dataset in an Analysis task.
2) We use `secondaryDataset` for a secondary dataset in an Analysis task.
3) We use `primaryDataset` for the first portion of the publication dataset name in a PrivateMC task.
I would prefer something like:
`inputDataset`
`secondaryInputDataset`
`publishPrimaryDataName`
and then we have still `publishDataName`.

**Comment by belforte** [comment]
In here, do you refer to names in code, or in crab configuration ?

**Comment by AndresTanasijczuk** [comment]
In CRAB configuration.

**Comment by belforte** [comment]
then I agree, but would use PrimaryDatasetName and publishDatasetName

**Comment by annawoodard** [comment]
I like the idea of changing `secondaryDataset` → `secondaryInputDataset`.

It is likely I'm missing something obvious here (sorry!), but what is the problem that changing `primaryDataset` is trying to fix? My vote would be to leave it as it is, because the name/use is analogous [1] to the use of primary datasets in central MC production, making it easier for people to understand what it is. And I don't see what extra information adding `publish` is; it doesn't strictly only affect publishing, it also affects the LFN, so it matters even if you have publishing disabled.

[1]
```
#central production (see https://twiki.cern.ch/twiki/bin/viewauth/CMS/DMWMPG_Namespace)
/store/mc/acquisition_era/primary-dataset/data_tier/processing_version/lfn_counter/filename.root

#CRAB3
/store/[user|group|local]/<dir>[/<subdirs>]/<primary-dataset>/<publication-name>/<time-stamp>/<counter>[/log]/<file-name>
```

**Comment by ericvaandering** [comment]
So @belforte "publishDatasetName" is yet another dataset related variable for completely overriding the generated dateset?

I like @annawoodard's suggestion.: secondaryInputDataset (or even auxiliary, which even I can't spell) How about outputDataset and outputPrimaryDataset for the two @belforte mentions? 

But the large question is: Is this really causing enough confusion that re-mapping things is warranted?

**Comment by AndresTanasijczuk** [comment]
Ok, seems we all agree to rename secondaryDataset to secondaryInputDataset (or auxiliary as per Eric).

About primaryDataset, maybe Anna is right that it can stay as it is. The confusion I wanted to solve is mainly about primaryDataset and secondaryDataset having nothing to do with each other but from their names sounding they refer to a same thing (input datasets) just one is the primary and the other the secondary.

**Comment by annawoodard** [comment]
I also like @ericvaandering's suggestion of auxiliary. I think @AndresTanasijczuk's suggestion of changing `secondaryDataset` already solves the problem of `primaryDataset` and `secondaryDataset` sounding too related, but I like @ericvaandering's suggestion to add 'output' for clarification. 

Sorry to go off on a tangent here, but I think using the word 'name' in a parameter name is usually (not always!) redundant. What's the difference between 'dataset' and 'datasetName'? If we're trying to express the fact that 'datasetName' is a piece of the full dataset string, then it is clearer in my opinion to either omit 'name', or use 'label' or 'tag' instead.

So, my vote is:

`inputDataset`→ `inputDataset`
`secondaryDataset` → `auxiliaryInputDataset`
`primaryDataset` → `primaryDataset` or `outputPrimaryDataset`
`publishDataName` → `dataset` or `outputDataset`

**Comment by belforte** [comment]
+1
thanks for pointing it out. We non-native speakers are often miselad

**Comment by ericvaandering** [comment]
Yeah. My problem is I'm stuck on what I would call primaryDataset since it's only part of a full dataset name and that's not reflected. Of course, this is a CMS problem, not just a CRAB one.

**Comment by belforte** [comment]
yup, but I am assuming we can't get DBS to change the way they decided to name things in the beginning

**Comment by annawoodard** [comment]
> Yeah. My problem is I'm stuck on what I would call primaryDataset since it's only part of a full dataset name and that's not reflected. Of course, this is a CMS problem, not just a CRAB one.

Yeah, I guess the best we can do is try to reduce the number of mappings where possible so that our nomenclature lines up as well as possible with central CMS nomenclature. In the case of changing `publishDataName` to `dataset` or `outputDataset`, @ericvaandering's point is even more relevant: those names are misleading because it's only a part of the dataset name and that's not reflected. On the other hand, the 'publish' in `publishDataName` is misleading because it is relevant even if publishing is disabled, and the 'name' is redundant as discussed earlier. So I amend my previous vote to:

`inputDataset`→ `inputDataset`
`secondaryDataset` → `auxiliaryInputDataset`
`primaryDataset` → `primaryDataset` | `outputPrimaryDataset`
`publishDataName` → `datasetLabel` | `outputDatasetLabel` | `datasetTag` | `outputDatasetTag`

**Comment by belforte** [comment]
thanks @annawoodard for terse summaries.
We have two external constraints:
1. DBS.
   I wish we had a place where dataset, fulldataset, processeddataset etc. are clearly defined, but could not find it. Maybe @yuyiguo ?
  IIUC the rule is
  `full dataset name = /<primaryDS>/<processedDS>/Tier
 and what we call `publishDataName` in CRAB currently is probably a piece of the `processedDS` (since we add hash, time stamp...) ... so I vote for `outputDatasetTag`

2. CMSSW
 is auxiliaryDataset really free for us to name as we like, or is this in the end something that we hand to CMSSW in a place where they call it "secondary" ? If so, using a name other then CMSSW one
would be confusing. And we should stick with secondaryInput

**Comment by annawoodard** [comment]
> CMSSW is auxiliaryDataset really free for us to name as we like, or is this in the end something that we hand to CMSSW in a place where they call it "secondary" ?

Good point, @belforte. In the CMSSW config they use `pool.source.fileNames` and `pool.source.secondaryFileNames`, so yes, maybe 'secondary' > 'auxiliary'. I amend my vote again:

`inputDataset`→ `inputDataset`
`secondaryDataset` → `secondaryInputDataset`
`primaryDataset` → `primaryDataset` | `outputPrimaryDataset`
`publishDataName` → `datasetLabel` | `outputDatasetLabel` | `datasetTag` | `outputDatasetTag`

**Comment by ericvaandering** [comment]
@annawoodard now I am confused. I thought that inputDataset, secondaryDataset, and publishDataName as they are currently do, in fact, have to be a full dataset name while primaryDataset is only part of name.

I suggested auxiliary for exactly Stefano's point. It's another dataset that gets used for different things. Maybe the parent, maybe grandparent, maybe a sibling or cousin if CMSSW allows such a thing.

**Comment by belforte** [comment]
inputDataset and secondaryDataset: yes
publishDataName: no
yeah.. confusing.. that's why we are having fun here :-)

**Comment by belforte** [comment]
hmm..
so I guess we can say
"for each job Crab finds the proper relative files in the auxiliary dateset and feeds them to CMSSW 
as pool.source.secondoryFileNames"
does not sound too bad. If anything, different people findind different naming more clear should 
mean that it is not a big deal either way.

**Comment by annawoodard** [comment]
> I thought that inputDataset, secondaryDataset, and publishDataName as they are currently do, in fact, have to be a full dataset name while primaryDataset is only part of name.

That's true. I think the choice of 'primary dataset' centrally is unfortunate :(.

> I suggested auxiliary for exactly Stefano's point. It's another dataset that gets used for different things. Maybe the parent, maybe grandparent, maybe a sibling or cousin if CMSSW allows such a thing.

Yeah, I was just thinking of how the mapping between crab project to cmsRun job goes, using the word that's used in pool source might make things clearer.
`inputDataset` → `pool.source.fileNames`
`secondaryInputDataset` → `pool.source.secondaryFileNames`

**Comment by belforte** [comment]
Original definitoin of DBS dataset name as /primary_ds/processed_ds/tier is here
https://twiki.cern.ch/twiki/bin/view/CMS/DBSNamingConvention#Naming_Conventions_for_Datasets

You can easily see that many things changed since then, but this stayed.
Along the way we sort of lost the idea of datasets as nested data containers and tend to
consider any string of the form /A/B/C as an independent entity, which makes Primary_ds and Processed_ds confusing names.

I still like Anna's proposal best:
'''
inputDataset→ inputDataset
secondaryDataset → secondaryInputDataset
primaryDataset → outputPrimaryDataset
publishDataName → outputDatasetTag
'''

**Comment by belforte** [comment]
and I still haven't learnt git markdown :-)

**Comment by ericvaandering** [comment]
So just to ease my confusion, A = outputPrimaryDataset and B=outputDatasetTag?

**Comment by belforte** [comment]
@ericvaandering I fail to parse your question.

A and B are parts of the final output dataset full name (aka path in DBS2 language)
see:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/Crab3DataHandling#Output_dataset_names_in_DBS

**Comment by ericvaandering** [comment]
My question is, are both
```
primaryDataset → outputPrimaryDataset
publishDataName → outputDatasetTag
```
only used to set part of the final output dataset name.

**Comment by yuyiguo** [comment]
@belforte full dataset name = /<primaryDS>/<processedDS>/Tier is what implemented in DBS. I knew there was a definition of dataset in DMWM twiki, but did not know exactly where it is. Maybe @hufnagel  knows?

Also processdDS is processed_ds_name=acquisition_era_name[-filter_name][-processing_str]-vprocessing_version where vprocessing_version is v and a number, such as v2.

Please see https://github.com/dmwm/DBS/blob/master/Server/Python/src/dbs/business/DBSBlockInsert.py#L618

**Comment by belforte** [comment]
@ericvaandering yes (also as part of the file's LFN, but that's sort of the same thing).

**Comment by AndresTanasijczuk** [comment]
I made a PR for changing `secondaryDataset` to `secondaryInputDataset`.
I didn't change the other parameters so far because I am not sure consensus was reached. @belforte I will wait for a final message from you.

**Comment by belforte** [comment]
@AndresTanasijczuk which PR ? sorry I am a bit confused at what you are exactly asking

**Comment by AndresTanasijczuk** [comment]
@belforte This PR: https://github.com/dmwm/CRABClient/pull/4562

**Comment by belforte** [comment]
ok., that's a good PR.
IIUC your question, you want to know about the full renaming:


inputDataset→ inputDataset
secondaryDataset → secondaryInputDataset
primaryDataset → outputPrimaryDataset
publishDataName → outputDatasetTag

I think we all agree and that can be done as well.

**Comment by AndresTanasijczuk** [comment]
Ok. I will then extend PR #4562 to include the other 2 changes (`inputDataset` is not changing).

**Comment by AndresTanasijczuk** [comment]
Done.

**Comment by AndresTanasijczuk** [comment]
Closing. Fixed in #4562
